File: bindings/rust/src/stream.rs
//! Safe `StreamEncoder` wrapper around `ztok_stream*`.
//!
//! Two surfaces:
//!
//! 1. `StreamEncoder::feed` / `finish` — sink-style API matching the
//!    Node `koffi` and Python ctypes shape. Returns each batch of
//!    newly-emitted ids as a `Vec<u32>`.
//!
//! 2. `Pipeline::encode_stream` + `StreamIter` — an `Iterator` adapter
//!    that yields one `Vec<u32>` per non-empty emit, including the
//!    final flush. Idiomatic Rust shape, useful in `for` loops.
//!
//! The encoder defers a trailing partial UTF-8 codepoint / pre-tokenizer
//! span up to a soft 1 MiB cap (see `src/stream.zig`); past that it
//! force-cuts at the nearest codepoint boundary.

use core::ffi::c_int;

use crate::batch::materialize_and_free;
use crate::error::{check_status, Error, Result};
use crate::pipeline::Pipeline;
use crate::sys;

/// Default feed chunk size (64 KiB) — matches the Python/Go bindings.
pub const DEFAULT_CHUNK_SIZE: usize = 64 * 1024;

/// A streaming-encode session against a [`Pipeline`].
///
/// Construct via [`StreamEncoder::new`] or [`Pipeline::encode_stream`].
/// Feed bytes via [`StreamEncoder::feed`] and call [`StreamEncoder::finish`]
/// to drain the final carry.
pub struct StreamEncoder<'a> {
    handle: *mut sys::ZtokStream,
    // Keep the pipeline alive for the encoder's lifetime. The stream
    // borrows from the pipeline (the C ABI documents that the stream
    // holds a pointer to the pipeline internally, see src/stream.zig),
    // so dropping the pipeline first would dangle.
    _pipeline: &'a Pipeline,
    finished: bool,
}

impl<'a> StreamEncoder<'a> {
    /// Open a new streaming session against `pipeline`.
    pub fn new(pipeline: &'a Pipeline) -> Result<Self> {
        let mut status: c_int = sys::ZTOK_OK;
        let h = unsafe { sys::ztok_stream_new(pipeline.raw(), &mut status) };
        check_status(status, "ztok_stream_new")?;
        if h.is_null() {
            return Err(Error::NullHandle {
                op: "ztok_stream_new",
            });
        }
        Ok(Self {
            handle: h,
            _pipeline: pipeline,
            finished: false,
        })
    }

    /// Feed `bytes` to the encoder. Returns any token ids newly emitted
    /// by this call (may be empty when the encoder is still buffering
    /// toward the next safe cut).
    pub fn feed(&mut self, bytes: &[u8]) -> Result<Vec<u32>> {
        if self.finished {
            return Err(Error::Internal {
                status: sys::ZTOK_ERR_INVALID_INPUT,
                op: "ztok_stream_feed: stream already finished",
            });
        }
        let mut out_ids: *mut u32 = core::ptr::null_mut();
        let mut out_n: usize = 0;
        let rc = unsafe {
            let (ptr, len) = if bytes.is_empty() {
                (core::ptr::null(), 0)
            } else {
                (bytes.as_ptr() as *const _, bytes.len())
            };
            sys::ztok_stream_feed(self.handle, ptr, len, &mut out_ids, &mut out_n)
        };
        if let Err(e) = check_status(rc, "ztok_stream_feed") {
            // Even on error, the C ABI may have allocated a partial
            // buffer — free it before propagating.
            if !out_ids.is_null() {
                unsafe { sys::ztok_ids_free(out_ids) };
            }
            return Err(e);
        }
        Ok(materialize_and_free(out_ids, out_n))
    }

    /// Flush any remaining carry as a final encode. Idempotent: a
    /// second call returns an empty `Vec`.
    pub fn finish(&mut self) -> Result<Vec<u32>> {
        if self.finished {
            return Ok(Vec::new());
        }
        self.finished = true;
        let mut out_ids: *mut u32 = core::ptr::null_mut();
        let mut out_n: usize = 0;
        let rc =
            unsafe { sys::ztok_stream_finish(self.handle, &mut out_ids, &mut out_n) };
        if let Err(e) = check_status(rc, "ztok_stream_finish") {
            if !out_ids.is_null() {
                unsafe { sys::ztok_ids_free(out_ids) };
            }
            return Err(e);
        }
        Ok(materialize_and_free(out_ids, out_n))
    }
}

impl<'a> Drop for StreamEncoder<'a> {
    fn drop(&mut self) {
        if !self.handle.is_null() {
            unsafe { sys::ztok_stream_free(self.handle) };
            self.handle = core::ptr::null_mut();
        }
    }
}

/// `Iterator` adapter over a [`StreamEncoder`] driven by a fixed input
/// buffer. Yields one non-empty `Vec<u32>` per emit; the final flush is
/// emitted last.
///
/// Returned by [`Pipeline::encode_stream`].
pub struct StreamIter<'a> {
    encoder: StreamEncoder<'a>,
    data: &'a [u8],
    cursor: usize,
    chunk_size: usize,
    drained: bool,
}

impl<'a> StreamIter<'a> {
    pub(crate) fn new(pipeline: &'a Pipeline, data: &'a [u8], chunk_size: usize) -> Result<Self> {
        Ok(Self {
            encoder: StreamEncoder::new(pipeline)?,
            data,
            cursor: 0,
            chunk_size: chunk_size.max(1),
            drained: false,
        })
    }
}

impl<'a> Iterator for StreamIter<'a> {
    type Item = Result<Vec<u32>>;

    fn next(&mut self) -> Option<Self::Item> {
        // Loop until we either produce a non-empty batch, hit the final
        // flush, or exhaust everything — the C ABI happily returns
        // empty emit lists between cuts, but exposing those would be
        // noise.
        loop {
            if self.cursor < self.data.len() {
                let end = (self.cursor + self.chunk_size).min(self.data.len());
                let chunk = &self.data[self.cursor..end];
                self.cursor = end;
                match self.encoder.feed(chunk) {
                    Ok(ids) if ids.is_empty() => continue,
                    Ok(ids) => return Some(Ok(ids)),
                    Err(e) => return Some(Err(e)),
                }
            }
            if !self.drained {
                self.drained = true;
                match self.encoder.finish() {
                    Ok(ids) if ids.is_empty() => return None,
                    Ok(ids) => return Some(Ok(ids)),
                    Err(e) => return Some(Err(e)),
                }
            }
            return None;
        }
    }
}

impl Pipeline {
    /// Stream-encode `data` and produce one `Vec<u32>` per non-empty
    /// emit. The final flush is yielded last. `chunk_size` controls the
    /// per-feed byte count; pass `0` (or use [`encode_stream_default`])
    /// to get the standard 64 KiB.
    ///
    /// [`encode_stream_default`]: Pipeline::encode_stream_default
    pub fn encode_stream<'a>(
        &'a self,
        data: &'a [u8],
        chunk_size: usize,
    ) -> Result<StreamIter<'a>> {
        StreamIter::new(self, data, chunk_size)
    }

    /// Convenience wrapper that uses [`DEFAULT_CHUNK_SIZE`].
    pub fn encode_stream_default<'a>(&'a self, data: &'a [u8]) -> Result<StreamIter<'a>> {
        self.encode_stream(data, DEFAULT_CHUNK_SIZE)
    }
}

File: src/stream.zig
//! Streaming encode API — emit ids as input bytes arrive, instead of
//! waiting for the full buffer.
//!
//! Use case: partial chat responses, stdin pipelines, long-document
//! pre-tokenization. The streaming encoder defers any trailing partial
//! codepoint OR partial pre-tokenizer span at the end of each `feed`
//! call to a small carry buffer; on the next `feed` it prepends that
//! carry and re-encodes from the resulting safe prefix.
//!
//! Safety model — `findSafeCut`-based:
//!   The pre-tokenizer's `findSafeCut(input, desired, window)` tells us
//!   the largest position `<= desired` at which `input` can be split
//!   without changing the resulting token sequence on either half.
//!   `feed` runs `findSafeCut(buf, buf.len, window)` on its accumulated
//!   buffer; bytes up to that cut point are encoded and emitted, and
//!   the remainder becomes the new carry.
//!
//! Bounded carry: the carry has a hard cap (`max_carry_bytes`, default
//! 1 MiB). If a single pre-tokenizer span — or a stretch of input with
//! no safe-cut boundary — exceeds this, the encoder force-encodes the
//! carry to recover. The output is still legal token-wise but MAY
//! diverge from single-shot `Pipeline.encode` at the forced cut (the
//! same divergence `encodeChunked` documents when no safe cut is
//! found within the search window).
//!
//! Parity with single-shot `Pipeline.encode`:
//!   When the pre-tokenizer's `findSafeCut` finds a real boundary
//!   (cl100k regex match, HF byte-level whitespace), the streaming
//!   output is bit-identical to `encode(full_input)`. With the
//!   `identity` pre-tokenizer + a merge-based model (BPE/Monster),
//!   chunk boundaries fall on codepoint boundaries but NOT on BPE
//!   merge boundaries, so the streaming output MAY differ at chunk
//!   seams. This mirrors the same caveat `Pipeline.encodeChunked`
//!   documents — same boundary algorithm, same trade-off.
//!
//! Not used by `Pipeline.encode` itself; this is a complementary API
//! aimed at the `ztok serve` `/encode_stream` route and any library
//! caller that wants id-as-it-arrives semantics.

const std = @import("std");
const Pipeline = @import("pipeline.zig").Pipeline;
const ScratchArena = @import("pipeline.zig").ScratchArena;
const TokenId = @import("token.zig").TokenId;
const Vocab = @import("vocab.zig").Vocab;
const Bpe = @import("bpe.zig").Bpe;

/// Default cap on the carry buffer. If a single span exceeds this we
/// force a cut at the end of the carry; see the module doc comment.
pub const default_max_carry_bytes: usize = 1 * 1024 * 1024;

/// Default search window for `findSafeCut`. Mirrors the
/// `encodeChunked` window (16 KiB is plenty to find a regex boundary
/// in the long-line corpora we've seen — max line length ~11 KB on
/// Wikipedia).
pub const default_search_window: usize = 16 * 1024;

pub const Options = struct {
    /// Hard cap on the carry buffer. The encoder returns
    /// `error.CarryOverflow` only if the underlying allocator fails —
    /// this cap is a soft limit: when exceeded, the encoder force-
    /// encodes the carry and emits the resulting ids. See module doc.
    max_carry_bytes: usize = default_max_carry_bytes,
    /// Window passed to `pre_tokenizer.findSafeCut`. Larger windows
    /// can find more cut points but cost O(window) per `feed`.
    search_window: usize = default_search_window,
};

/// Persistent streaming-encode state. One per stream.
///
/// `feed(bytes, out)` appends to the carry, then encodes everything up
/// to the nearest safe cut, appending ids to `out` (any type that has
/// `append(allocator, TokenId)` and `appendSlice(allocator, []const
/// TokenId)` — `*std.ArrayList(TokenId)` is the obvious choice).
///
/// `finish(out)` flushes the remaining carry as a final encode.
///
/// The same allocator backs the carry and the internal scratch arena.
/// The caller owns whatever output container they pass in.
pub const StreamEncoder = struct {
    allocator: std.mem.Allocator,
    pipeline: *const Pipeline,
    scratch: ScratchArena,
    carry: std.ArrayList(u8),
    opts: Options,

    pub fn init(
        allocator: std.mem.Allocator,
        pipeline: *const Pipeline,
    ) StreamEncoder {
        return initWithOptions(allocator, pipeline, .{});
    }

    pub fn initWithOptions(
        allocator: std.mem.Allocator,
        pipeline: *const Pipeline,
        opts: Options,
    ) StreamEncoder {
        return .{
            .allocator = allocator,
            .pipeline = pipeline,
            .scratch = ScratchArena.init(allocator),
            .carry = .empty,
            .opts = opts,
        };
    }

    pub fn deinit(self: *StreamEncoder) void {
        self.carry.deinit(self.allocator);
        self.scratch.deinit();
    }

    /// Bytes currently buffered (not yet emitted).
    pub fn pendingBytes(self: *const StreamEncoder) usize {
        return self.carry.items.len;
    }

    /// Feed `bytes` to the encoder. Any ids that can now be safely
    /// emitted are appended to `out` via `out.append`/`out.appendSlice`.
    /// `out` must be `*std.ArrayList(TokenId)` (or anything with the
    /// same `append`/`appendSlice` shape — duck-typed).
    pub fn feed(
        self: *StreamEncoder,
        bytes: []const u8,
        out: anytype,
    ) !void {
        if (bytes.len == 0) return;
        try self.carry.appendSlice(self.allocator, bytes);
        try self.tryDrain(out, false);
    }

    /// Drop any held bytes into a final encode and emit them. Idempotent:
    /// callable twice in a row, second call is a no-op.
    pub fn finish(
        self: *StreamEncoder,
        out: anytype,
    ) !void {
        try self.tryDrain(out, true);
    }

    /// Encode `self.carry[0..cut]` through the pipeline, emit the ids,
    /// and shift `self.carry[cut..]` to the front of the carry.
    fn emitPrefix(
        self: *StreamEncoder,
        out: anytype,
        cut: usize,
    ) !void {
        if (cut == 0) return;
        const prefix = self.carry.items[0..cut];
        const ids = try self.pipeline.encodeWithScratch(self.allocator, prefix, &self.scratch);
        defer self.allocator.free(ids);
        try out.appendSlice(self.allocator, ids);

        const remaining = self.carry.items.len - cut;
        if (remaining > 0) {
            // Shift carry[cut..] to the front.
            std.mem.copyForwards(u8, self.carry.items[0..remaining], self.carry.items[cut..]);
        }
        self.carry.shrinkRetainingCapacity(remaining);
    }

    /// Drain emittable bytes. When `is_final` is true, flush everything
    /// remaining as a final encode.
    fn tryDrain(
        self: *StreamEncoder,
        out: anytype,
        is_final: bool,
    ) !void {
        if (is_final) {
            // Final flush: encode the entire carry, regardless of safe
            // cuts (this is the natural end-of-stream — any partial
            // codepoint left here is a bug in the upstream feeder, and
            // the model's UTF-8 handling will absorb it as best it can).
            if (self.carry.items.len > 0) {
                try self.emitPrefix(out, self.carry.items.len);
            }
            return;
        }

        if (self.carry.items.len == 0) return;

        // Find the largest safe cut <= carry.len. We never emit the very
        // last byte (it may be the start of a span that's still growing).
        // Search window is bounded by `opts.search_window`.
        const carry_len = self.carry.items.len;
        // `desired = carry_len - 1` searches backwards from the position
        // just before the last byte. findSafeCut will snap to the nearest
        // boundary <= desired.
        const desired = if (carry_len > 0) carry_len - 1 else 0;
        const maybe_cut = self.pipeline.pre_tokenizer.findSafeCut(
            self.carry.items,
            desired,
            self.opts.search_window,
        );

        var cut: usize = 0;
        if (maybe_cut) |c| {
            cut = c;
        } else if (carry_len >= self.opts.max_carry_bytes) {
            // Carry has grown past its soft cap and we still couldn't find
            // a safe cut in `search_window` bytes. Force a UTF-8-codepoint-
            // safe cut at the very end so we don't unbounded-grow.
            var p = carry_len;
            while (p > 0 and (self.carry.items[p - 1] & 0xC0) == 0x80) p -= 1;
            // back off the lead byte too if we're sitting mid-codepoint
            if (p > 0 and (self.carry.items[p - 1] & 0xC0) == 0xC0) p -= 1;
            cut = p;
        }
        if (cut > 0) try self.emitPrefix(out, cut);
    }
};

// === Tests ============================================================

const Normalizer = @import("normalizer.zig").Normalizer;
const PreTokenizer = @import("pretok.zig").PreTokenizer;
const Model = @import("model.zig").Model;
const Decoder = @import("decoder.zig").Decoder;
const added_tokens_mod = @import("added_tokens.zig");

test "StreamEncoder: 3-chunk feed matches single-shot encode (cl100k+bpe)" {
    const a = std.testing.allocator;

    // Build a small cl100k-style BPE vocab matching the existing
    // pipeline test: 256 byte ids + a few extra merges so "hello",
    // " world" tokenize to a couple of ids.
    var src: std.ArrayList(u8) = .empty;
    defer src.deinit(a);
    const b64 = std.base64.standard.Encoder;
    var enc_buf: [16]u8 = undefined;
    var rank: u32 = 0;
    var b: u32 = 0;
    while (b < 256) : (b += 1) {
        const byte: [1]u8 = .{@intCast(b)};
        const encoded = b64.encode(&enc_buf, &byte);
        try src.print(a, "{s} {d}\n", .{ encoded, rank });
        rank += 1;
    }
    const extra = [_][]const u8{
        "he",     "hel",    "hell",   "hello",
        " w",     " wo",    " wor",   " worl",
        " world", " q",     " qu",    " qui",
        " quic",  " quick", " b",     " br",
        " bro",   " brow",  " brown", " f",
        " fo",    " fox",
    };
    for (extra) |bytes| {
        const encoded = b64.encode(&enc_buf, bytes);
        try src.print(a, "{s} {d}\n", .{ encoded, rank });
        rank += 1;
    }

    var bpe = try Bpe.loadTiktokenBytes(a, src.items);
    defer bpe.deinit();

    var v = Vocab.empty(a);
    defer v.deinit();

    const pipe: Pipeline = .{
        .normalizer = .identity,
        .pre_tokenizer = .cl100k,
        .model = .{ .bpe = &bpe },
        .decoder = .concat,
        .vocab = &v,
    };

    const full = "hello world the quick brown fox";
    const want = try pipe.encode(a, full);
    defer a.free(want);

    // Feed in three deliberately uneven chunks.
    var enc = StreamEncoder.init(a, &pipe);
    defer enc.deinit();
    var got: std.ArrayList(TokenId) = .empty;
    defer got.deinit(a);

    try enc.feed(full[0..11], &got); // "hello world"
    try enc.feed(full[11..21], &got); // " the quick"
    try enc.feed(full[21..], &got); // " brown fox"
    try enc.finish(&got);

    try std.testing.expectEqualSlices(TokenId, want, got.items);
}

test "StreamEncoder: mid-UTF-8 codepoint bytes do not produce garbage ids" {
    const a = std.testing.allocator;

    // byte_id model so we can predict ids exactly.
    var v = Vocab.empty(a);
    defer v.deinit();
    const pipe: Pipeline = .{
        .normalizer = .identity,
        .pre_tokenizer = .identity,
        .model = .byte_id,
        .decoder = .concat,
        .vocab = &v,
    };

    // "héllo" — the é is a 2-byte UTF-8 sequence (0xC3 0xA9). We feed
    // up to the first byte of é, then the rest.
    const full = "h\xC3\xA9llo";
    const want = try pipe.encode(a, full);
    defer a.free(want);

    var enc = StreamEncoder.init(a, &pipe);
    defer enc.deinit();
    var got: std.ArrayList(TokenId) = .empty;
    defer got.deinit(a);

    // Feed everything up to and INCLUDING the leading byte of the é.
    try enc.feed(full[0..2], &got);
    // The half-codepoint must NOT have been emitted yet. byte_id with
    // identity pretok will emit at most one byte of safe content (the
    // 'h') before the carry; the trailing 0xC3 must be deferred.
    // We assert nothing past 'h' has been emitted.
    try std.testing.expect(got.items.len <= 1);
    if (got.items.len == 1) {
        try std.testing.expectEqual(@as(TokenId, 'h'), got.items[0]);
    }

    try enc.feed(full[2..], &got);
    try enc.finish(&got);
    try std.testing.expectEqualSlices(TokenId, want, got.items);
}

test "StreamEncoder: empty feed + finish produces zero ids" {
    const a = std.testing.allocator;

    var v = Vocab.empty(a);
    defer v.deinit();
    const pipe: Pipeline = .{
        .normalizer = .identity,
        .pre_tokenizer = .identity,
        .model = .byte_id,
        .decoder = .concat,
        .vocab = &v,
    };

    var enc = StreamEncoder.init(a, &pipe);
    defer enc.deinit();
    var got: std.ArrayList(TokenId) = .empty;
    defer got.deinit(a);

    try enc.feed("", &got);
    try enc.finish(&got);
    try std.testing.expectEqual(@as(usize, 0), got.items.len);

    // Idempotent finish.
    try enc.finish(&got);
    try std.testing.expectEqual(@as(usize, 0), got.items.len);
}

test "StreamEncoder: many-tiny-chunks matches single-shot" {
    const a = std.testing.allocator;
    var v = Vocab.empty(a);
    defer v.deinit();
    const pipe: Pipeline = .{
        .normalizer = .identity,
        .pre_tokenizer = .identity,
        .model = .byte_id,
        .decoder = .concat,
        .vocab = &v,
    };

    const full = "the quick brown fox jumps over the lazy dog";
    const want = try pipe.encode(a, full);
    defer a.free(want);

    var enc = StreamEncoder.init(a, &pipe);
    defer enc.deinit();
    var got: std.ArrayList(TokenId) = .empty;
    defer got.deinit(a);

    // Feed one byte at a time.
    for (full) |c| {
        const one = [_]u8{c};
        try enc.feed(&one, &got);
    }
    try enc.finish(&got);
    try std.testing.expectEqualSlices(TokenId, want, got.items);
}
